src/MongoDB.Driver.Core/Core/Authentication/ScramCache.cs

```diff
@@ -0,0 +1,144 @@
+* limitations under the License.
+*/
+
+using System;
+using System.Linq;
+using System.Security;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Shared;
+
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// A cache for Client and Server keys, to be used during authentication.
+    /// </summary>
+    internal class ScramCache
+    {
+        private ScramCacheKey _cacheKey;
+        private ScramCacheEntry _cachedEntry;
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="key"></param>
+        /// <param name="entry"></param>
+        /// <returns></returns>
+        public bool TryGet(ScramCacheKey key, out ScramCacheEntry entry)
+        {
+            if (_cacheKey.Equals(key))
+            {
+                entry = _cachedEntry;
+                return true;
+            }
+            else
+            {
+                entry = null;
+                return false;
+            }
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="key"></param>
+        /// <param name="entry"></param>
+        public void Set(ScramCacheKey key, ScramCacheEntry entry)
+        {
+            _cacheKey = key;
+            _cachedEntry = entry;
+        }
+    }
+
+    internal class ScramCacheKey
+    {
+        private int _iterationCount;
+        private SecureString _password;
+        private byte[] _salt;
+
+        internal ScramCacheKey(SecureString password, byte[] salt, int iterationCount)
+        {
+            _iterationCount = iterationCount;
+            _password = password;
+            _salt = salt;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (this == obj)
+            {
+                return true;
+            }
+
+            if (obj == null || obj.GetType() != obj.GetType())
+            {
+                return false;
+            }
+
+            ScramCacheKey other = (ScramCacheKey) obj;
+
+            return
+                Equals(_password,other._password) &&
+                _iterationCount == other._iterationCount &&
+                _salt.SequenceEqual(other._salt);
+        }
+
+        // TODO: still need to find a correct way to get hash of byte[]
+        public override int GetHashCode()
+        {
+            int hash = 17;
+            hash = 37 * hash + _iterationCount.GetHashCode();
+            hash = 37 * hash + GetHashCode(_password);
+            hash = 37 * hash + _salt.GetHashCode();
+            return hash;
+        }
+
+        // private methods
+        private bool Equals(SecureString x, SecureString y)
+        {
+            using (var dx = new DecryptedSecureString(x))
+            using (var dy = new DecryptedSecureString(y))
+            {
+                var xchars = dx.GetChars();
+                var ychars = dy.GetChars();
+                return xchars.SequenceEqual(ychars);
```

Comment:
Turns out SequenceEqual is rather inefficient. Let's replace SequenceEqual with

    return Equals(xchars, ychars);

with this more efficient helper method for comparing char arrays:

    private bool Equals(char[] x, char[] y)
    {
        if (x.Length != y.Length)
        {
            return false;
        }

        for (var i = 0; i < x.Length; i++)
        {
            if (x[i] != y[i])
            {
                return false;
            }
        }

        return true;
    }

Reply:
Done

Reply:
Wow, SequenceEquals is really quite slow thanks to it using IEnumerator and having to box and unbox... good to know! 🤔 😄